File: 016.NNCoreAnimation/NNCoreAnimation/Controllers/NNRotationViewController.m
//
//  NNRotationViewController.m
//  NNCoreAnimation
//
//  Created by 刘朋坤 on 2018/8/13.
//  Copyright © 2018年 刘朋坤. All rights reserved.
//

#import "NNRotationViewController.h"

@interface NNRotationViewController ()

@end

@implementation NNRotationViewController

- (NSArray *)buttonTitleArray {
    return @[@"rotation.x", @"rotation.y", @"rotation.z", @"组合"];
}

- (void)buttonClick:(UIButton *)sender {
    switch (sender.tag) {
        case 0:
            [self rotationAnimationWithKeyPath:@"transform.rotation.x"];
            break;
        case 1:
            [self rotationAnimationWithKeyPath:@"transform.rotation.y"];
            break;
        case 2:
            [self rotationAnimationWithKeyPath:@"transform.rotation.z"];
            break;
        case 3:
            [self groupAnimation];
            break;
        default:
            break;
    }
}

- (void)rotationAnimationWithKeyPath:(NSString *)KeyPath {
    CABasicAnimation *anima = [CABasicAnimation animationWithKeyPath:KeyPath];
    anima.toValue = [NSNumber numberWithFloat:M_PI];
    anima.duration = 2.0f;
    [self.nnView.headerImage.layer addAnimation:anima forKey:@"rotateAnimation"];
}

- (void)groupAnimation {
    CFTimeInterval currentTime = CACurrentMediaTime();
    // 位移动画
    CAKeyframeAnimation *animation1 = [CAKeyframeAnimation animationWithKeyPath:@"position"];
    NSValue *value0 = [NSValue valueWithCGPoint:CGPointMake(100, 100)];
    NSValue *value1 = [NSValue valueWithCGPoint:CGPointMake(200, 300)];
    NSValue *value2 = [NSValue valueWithCGPoint:CGPointMake(500, 100)];
    NSValue *value3 = [NSValue valueWithCGPoint:CGPointMake(100, 500)];
    NSValue *value4 = [NSValue valueWithCGPoint:CGPointMake(300, 300)];
    animation1.values = [NSArray arrayWithObjects:value0, value1, value2, value3, value4, nil];
    animation1.beginTime = currentTime;
    animation1.duration = 2.0f;
    animation1.fillMode = kCAFillModeForwards;
    animation1.removedOnCompletion = NO;
    [self.nnView.headerImage.layer addAnimation:animation1 forKey:@"position"];
    
    // 缩放动画
    CABasicAnimation *animation2 = [CABasicAnimation animationWithKeyPath:@"transform.scale"];
    animation2.fromValue = [NSNumber numberWithFloat:0.8f];
    animation2.toValue = [NSNumber numberWithFloat:1.5f];
    animation2.beginTime = currentTime + 2.0f;
    animation2.duration = 1.0f;
    animation2.fillMode = kCAFillModeForwards;
    animation2.removedOnCompletion = NO;
    [self.nnView.headerImage.layer addAnimation:animation2 forKey:@"transform.scale"];
    
    // 旋转动画
    CABasicAnimation *animation3 = [CABasicAnimation animationWithKeyPath:@"transform.rotation"];
    animation3.toValue = [NSNumber numberWithFloat:M_PI * 4];
    animation3.beginTime = currentTime + 3.0f;
    animation3.duration = 1.0f;
    animation3.fillMode = kCAFillModeForwards;
    animation3.removedOnCompletion = NO;
    [self.nnView.headerImage.layer addAnimation:animation3 forKey:@"transform.rotation"];
}

@end

File: 018.NNCoreAnimation_swift/NNCoreAnimation_swift/NNCoreAnimation_swift/Controllers/NNRotationViewController.swift
//
//  NNRotationViewController.swift
//  NNCoreAnimation_swift
//
//  Created by 刘朋坤 on 2019/1/5.
//  Copyright © 2019 刘朋坤. All rights reserved.
//

import UIKit

class NNRotationViewController: NNBaseViewController {

    // 按钮点击事件
    override func buttonClick(index: NSInteger) {
        switch index {
        case 0:
            rotationAnimationWithKeyPath(KeyPath: "transform.rotation.x")
            break
        case 1:
            rotationAnimationWithKeyPath(KeyPath: "transform.rotation.y")
            break
        case 2:
            rotationAnimationWithKeyPath(KeyPath: "transform.rotation.z")
            break
        case 3:
            groupAnimation()
            break
        default:
            break
        }
    }
    
    // 封装的统一方法
    func rotationAnimationWithKeyPath(KeyPath:String) {
        let anima = CABasicAnimation.init(keyPath: KeyPath)
        anima.toValue = M_PI
        anima.duration = 2
        self.nnView.headerImage.layer.add(anima, forKey: "rotateAnimation")
    }
    
    // 组合动画
    func groupAnimation() {
        let currentTime = CACurrentMediaTime()
        
        // 位移动画
        let animation1 = CAKeyframeAnimation.init(keyPath: "position")
        let value1 = NSValue.init(cgPoint: CGPoint.init(x: 100, y: 100))
        let value2 = NSValue.init(cgPoint: CGPoint.init(x: 200, y: 300))
        let value3 = NSValue.init(cgPoint: CGPoint.init(x: 500, y: 100))
        let value4 = NSValue.init(cgPoint: CGPoint.init(x: 100, y: 500))
        let value5 = NSValue.init(cgPoint: CGPoint.init(x: 300, y: 100))
        animation1.values = [value1, value2, value3, value4, value5]
        animation1.beginTime = currentTime
        animation1.duration = 2
        animation1.fillMode = .forwards
        animation1.isRemovedOnCompletion = false
        self.nnView.headerImage.layer.add(animation1, forKey: "position")
        
        // 缩放动画
        let animation2 = CABasicAnimation.init(keyPath: "transform.scale")
        animation2.fromValue = 0.8
        animation2.toValue = 1.5
        animation2.beginTime = currentTime + 2
        animation2.duration = 1
        animation2.fillMode = .forwards
        animation2.isRemovedOnCompletion = false
        self.nnView.headerImage.layer.add(animation2, forKey: "transform.scale")
        
        // 旋转动画
        let animation3 = CABasicAnimation.init(keyPath: "transform.rotation")
        animation3.toValue = .pi * 4.0
        animation3.beginTime = currentTime + 3
        animation3.duration = 1
        animation3.fillMode = .forwards
        animation3.isRemovedOnCompletion = false
        self.nnView.headerImage.layer.add(animation3, forKey: "transform.rotation")
    }
    
    // 数组元素
    override func buttonTitleArrayMethod() -> NSArray {
        return ["rotation.x", "rotation.y", "rotation.z", "组合"]
    }

}
